crates/exex/exex/src/manager.rs

```diff
@@ -303,83 +312,152 @@ impl ExExManager {
+            .map(|exex_handle| {
+                (&exex_handle.id, exex_handle.finished_height.map(|block| block.hash))
+            })
+            // Deduplicate all hashes
+            .unique_by(|(_, hash)| *hash)
+            // Check if hashes are canonical
+            .map(|(exex_id, hash)| {
+                hash.map_or(Ok((exex_id, hash, false)), |hash| {
+                    self.provider
+                        .is_known(&hash)
+                        // Save the ExEx ID, hash of the finished height, and whether the hash
+                        // is canonical
+                        .map(|is_canonical| (exex_id, Some(hash), is_canonical))
+                })
+            })
+            // We collect here to be able to log the unfinalized ExExes below
+            .collect::<Result<Vec<_>, _>>()?;
+        if exex_finished_heights.iter().all(|(_, _, is_canonical)| *is_canonical) {
+            // If there is a finalized header and all ExExs are on the canonical chain, finalize
+            // the WAL with the new finalized header
+            self.wal.finalize(finalized_header.num_hash())?;
+        } else {
+            let unfinalized_exexes = exex_finished_heights
+                .into_iter()
+                .filter_map(|(exex_id, hash, is_canonical)| {
+                    is_canonical.not().then_some((exex_id, hash))
+                })
+                .collect::<Vec<_>>();
+            debug!(
+                ?unfinalized_exexes,
+                "Not all ExExes are on the canonical chain, can't finalize the WAL"
+            );
+        }
+
+        Ok(())
+    }
+}
+
+impl<P> Future for ExExManager<P>
+where
+    P: HeaderProvider + Unpin + 'static,
+{
     type Output = eyre::Result<()>;
 
-    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
-        // Drain the finalized header stream and grab the last finalized header
+    /// Main loop of the [`ExExManager`]. The order of operations is as follows:
+    /// 1. Handle incoming ExEx events. We do it before finalizing the WAL, because it depends on
+    ///    the latest state of [`ExExEvent::FinishedHeight`] events.
+    /// 2. Finalize the WAL with the finalized header, if necessary.
+    /// 3. Drain [`ExExManagerHandle`] notifications, push them to the internal buffer and update
+    ///    the internal buffer capacity.
+    /// 5. Send notifications from the internal buffer to those ExExes that are ready to receive new
+    ///    notifications.
+    /// 5. Remove notifications from the internal buffer that have been sent to **all** ExExes and
+    ///    update the internal buffer capacity.
+    /// 6. Update the channel with the lowest [`FinishedExExHeight`] among all ExExes.
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.get_mut();
+
+        // Handle incoming ExEx events
+        for exex in &mut this.exex_handles {
```

Comment:
do we actually spawn the manager if there are 0 exexs?

if we do, then we need a shortcut because otherwise we do unnecessary wal IO?

Reply:
we don't spawn it if there are no ExExes